doc/source/object-store.rst

```diff
@@ -0,0 +1,169 @@
+Object Store (Plasma)
+=====================
+
+Plasma is an in-memory object store that is being developed as part of `Apache Arrow`_. Ray uses Plasma to efficiently transfer objects across different processes and different nodes.
+
+All objects in Plasma object store are **immutable** and held in shared memory. This is so that they can be accessed efficiently by many workers on the same node.
+
+Note that Plasma does not support all data-types, and in certain cases it may be necessary to either write your own serialization protocol or use Actors to hold objects and transfer object state (i.e., weight matrices) among Ray workers.
+
+Overview
+--------
+
+Ray will place objects in a node object store in the following situations:
+
+1. Calling ``ray.put``
+
+.. code:: python
+
+    y = 2
+    # This places `2` into the object store.
+    object_id = ray.put(y)
+
+2. The return values of a remote function.
+
+.. code:: python
+
+    @ray.remote
+    def remote_function():
+        return 1
+
+    # This places `1` into the object store.
+    object_id = remote_function.remote()
+
+
+3. Arguments to remote functions (except for simple arguments like ints or
+   floats).
+
+.. code:: python
+
+    @ray.remote
+    def remote_function(y):
+        # Note that inside the remote function, the actual argument is provided.
+        return len(y)
+
+    argument = [1, 2, 3, 4]
+    # This implicitly places `argument` into the object store.
+    remote_function.remote(argument)
+
+In the single node setting, all Ray workers can read the same object in the object store without copying (zero-copy reads). Once an object is placed in the object store, it is immutable. Any writes to the read-only object will result in a copy into the local process memory.
+
+Distributed Setting
+~~~~~~~~~~~~~~~~~~~
+
+Each node has its own object store. When data is put into a node object store, it
+does not get automatically broadcasted; it remains local until requested by another
+task on another node.
+
+
+Numpy Arrays
+------------
+
+Ray optimizes for numpy arrays by using the `Apache Arrow`_ data format.
+Each numpy array object holds a pointer to the relevant array held in shared memory.
+
+There are some advantages to this form of serialization:
+
+- Deserialization can be very fast.
+- Memory is shared between processes so worker processes can all read the same
+  data without having to copy it.
+
+**Nested Numpy Arrays:** When we deserialize a list of numpy arrays from the object store, we still create a Python list of numpy array objects. However, rather than copy each numpy array,
```

Comment:
this sentence ends abruptly. Also, I'm not even sure what a nested numpy array is.